File: Linked List/linked_list.py
# Implementation of Singly Linked List in Python
'''
    This implementation defines a Node class to represent a node in the linked list, and a LinkedList 
    class to represent the linked list itself. The LinkedList class has methods to insert nodes at the 
    beginning or end of the list, delete nodes from the list, and print the list. Each node has two 
    attributes: its data and a reference to the next node in the list (next). 
    The LinkedList class has one attribute: the head node of the list (head). When a new node is 
    inserted at the beginning of the list, its next attribute is set to the current head node, and the 
    head attribute is set to the new node. When a new node is inserted at the end of the list, it is 
    appended to the last node's next attribute. When a node is deleted, the list is traversed to find 
    the node with the given data, and its next attribute is set to the node after it. Finally, the 
    print_list method traverses the list and prints each node's data.
'''
# Define a class to represent a node in the linked list
class Node:
    def __init__(self, data):
        # Each node has two attributes: data and a reference to the next node
        self.data = data
        self.next = None

# Define a class to represent the linked list itself
class LinkedList:
    def __init__(self):
        # Each linked list has one attribute: the head node (which initially is None)
        self.head = None
    
    # Method to insert a new node at the beginning of the linked list
    def insert_at_beginning(self, data):
        # Create a new node
        new_node = Node(data)
        
        # Set the next node of the new node to be the current head node
        new_node.next = self.head
        
        # Set the head node to be the new node
        self.head = new_node
    
    # Method to insert a new node at the end of the linked list
    def insert_at_end(self, data):
        # Create a new node
        new_node = Node(data)
        
        # If the list is empty, set the head node to be the new node
        if self.head is None:
            self.head = new_node
            return
        
        # Traverse to the last node in the list
        current_node = self.head
        while current_node.next:
            current_node = current_node.next
        
        # Set the next node of the last node to be the new node
        current_node.next = new_node
    
    # Method to delete the first occurrence of a node with the given data
    def delete_node(self, data):
        # If the list is empty, do nothing
        if self.head is None:
            return
        
        # If the node to be deleted is the head node, set the head node to be the next node
        if self.head.data == data:
            self.head = self.head.next
            return
        
        # Traverse the list to find the node to be deleted
        current_node = self.head
        while current_node.next:
            if current_node.next.data == data:
                current_node.next = current_node.next.next
                return
            current_node = current_node.next
    
    # Method to print the linked list
    def print_list(self):
        # Traverse the list and print each node's data
        current_node = self.head
        while current_node:
            print(current_node.data, end=" -> ")
            current_node = current_node.next
        print("None")

File: Linked List/LinkedList.java
package Linear.LinkedLists;

/**
 * Problem Description
 * Design and implement a Linked List data structure.
 * A node in a linked list should have the following attributes - an integer value and a pointer to the next node. It should support the following operations:
 *
 * insert_node(position, value) - To insert the input value at the given position in the linked list.
 * delete_node(position) - Delete the value at the given position from the linked list.
 * print_ll() - Print the entire linked list, such that each element is followed by a single space (no trailing spaces).
 * Note:
 *
 * If an input position does not satisfy the constraint, no action is required.
 * Each print query has to be executed in a new line.
 *
 *
 * Problem Constraints
 * 1 <= value <= 105
 * 1 <= position <= n where, n is the size of the linked-list.
 *
 *
 *
 * Input Format
 * First line contains an integer denoting number of cases, let's say t.
 * Next t line denotes the cases.
 *
 *
 * Output Format
 * When there is a case of print_ll(),  Print the entire linked list, such that each element is followed by a single space. There should not be any trailing space.
 * NOTE: You don't need to return anything.
 *
 *
 * Example Input
 * 5
 * i 1 23
 * i 2 24
 * p
 * d 1
 * p
 *
 *
 * Example Output
 * 23 24
 * 24
 *
 *
 * Example Explanation
 * After first two cases linked list contains two elements 23 and 24.
 * At third case print: 23 24.
 * At fourth case delete value at first position, only one element left 24.
 * At fifth case print: 24.
 */

public class LinkedList {
    static ListNode head;
    static int length;
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() { }
        ListNode(int val) {
            this.val = val;
        }
        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
    public static void main(String[] args) {

    }

    public static void insertNode(int position, int value) {
        if (position > length + 1) { // more than tail length
            return;
        }
        ListNode newNode = new ListNode(value);
        if (position == 1) {
            newNode.next = head;
            head = newNode;
        } else {
            ListNode prev = null;
            ListNode curr = head;
            int counter = 1;
            while (counter < position && curr != null) {
                prev = curr;
                curr = curr.next;
                counter++;
            }
            assert prev != null;
            prev.next = newNode;
            newNode.next = curr;
        }
        length++;
    }

    public static void deleteNode(int position) {
        if (position > length) {
            return;
        }

        if (position == 1) {
            head = head.next;
        } else {
            ListNode prev = null;
            ListNode curr = head;
            int counter = 1;
            while (counter < position && curr != null) {
                prev = curr;
                curr = curr.next;
                counter++;
            }
            assert prev != null;
            assert curr != null;
            prev.next = curr.next;
        }
        length--;
    }
    public static void printLinkedList() {
        if (length == 0) return;

        ListNode curr = head;

        while (curr != null) {
            if (curr.next != null)
                System.out.print(curr.val + " ");
            else System.out.print(curr.val);
            curr = curr.next;
        }
    }

}